deps/v8/src/api/api.cc

```diff
@@ -11255,37 +11255,65 @@ void RecordReplayInstrument(const char* kind, const char* function, int offset)
 extern char* CommandCallback(const char* command, const char* params);
 extern void ClearPauseDataCallback();
 
-bool gRecordReplayInstrumentNodeInternals;
-bool gRecordReplayAssertValues;
+// Whether execution within node internal scripts should advance the progress counter.
+static bool gRecordReplayInstrumentNodeInternals;
 
-bool ShouldEmitRecordReplayAssertValue() {
-  return gRecordReplayAssertValues;
-}
+// Whether JS assertions for execution and/or values are enabled.
+// This is not static so that it can be tested directly, for efficiency.
+bool gRecordReplayJSAsserts;
+
+// If non-zero, only execution asserts are enabled and only this frequency of
+// execution is asserted.
+static size_t gRecordRelayJSAssertFrequency;
 
-static std::vector<std::string> gRecordReplayAssertFilters;
+// Any source filters where JS asserts should be added.
+static std::vector<std::string> gRecordReplayJSAssertFilters;
 
-// The assertion filters are a comma separated list of patterns for sources where
-// record/replay asserts should be added.
-static void ProcessRecordReplayAssertFilters(const char* value) {
-  if (!value) {
+static void InitializeRecordReplayAsserts() {
+  gRecordReplayJSAsserts = !!getenv("RECORD_REPLAY_JS_ASSERTS");
+  if (!gRecordReplayJSAsserts) {
     return;
   }
-  while (true) {
-    const char* comma = strchr(value, ',');
-    std::string filter(value, comma ? comma - value : strlen(value));
-    gRecordReplayAssertFilters.push_back(filter);
-    if (!comma) {
-      break;
+
+  const char* frequency = getenv("RECORD_REPLAY_JS_ASSERT_FREQUENCY");
+  if (frequency) {
+    gRecordRelayJSAssertFrequency = atoi(frequency);
+  }
+
+  // The assertion filters are a comma separated list of patterns for sources where
+  // record/replay asserts should be added.
+  const char* filters = getenv("RECORD_REPLAY_JS_ASSERT_FILTERS");
+  if (filters) {
+    while (true) {
+      const char* comma = strchr(filters, ',');
+      std::string filter(filters, comma ? comma - filters : strlen(filters));
+      gRecordReplayJSAssertFilters.push_back(filter);
+      if (!comma) {
+        break;
+      }
+      filters = comma + 1;
     }
-    value = comma + 1;
   }
 }
 
+bool RecordReplayInstrumentNodeInternals() {
+  return gRecordReplayInstrumentNodeInternals;
+}
+
+// Whether to assert values seen during JS execution.
+bool RecordReplayAssertJSValues() {
+  return gRecordReplayJSAsserts && !gRecordRelayJSAssertFrequency;
+}
+
+bool RecordReplayShouldAssertForProgress(uint64_t progress) {
+  return !gRecordRelayJSAssertFrequency || !(progress % gRecordRelayJSAssertFrequency);
```

Comment:
Can the progress here increase by more that `1` at a time? What happens if we increment too much and skip the point where this check passes?

Reply:
Progress only ever increases by one at a time, but it can increase at places other than as JS execution advances, and we could skip points where we would otherwise add these asserts.  This won't cause problems though, it just means we'll have slightly fewer assertions sprinkled around in the application's execution.